modules/core/src/main/java/org/apache/ignite/internal/processors/cache/datastructures/CacheDataStructuresManager.java

```diff
@@ -421,29 +425,46 @@ public void onPartitionEvicted(int part) {
         try {
             GridCacheSetHeaderKey key = new GridCacheSetHeaderKey(name);
 
-            GridCacheSetHeader hdr;
+            boolean oldVer = true;
+
+            GridCacheSetHeader hdr = null;
 
             IgniteInternalCache cache = cctx.cache().withNoRetries();
 
-            if (create) {
-                hdr = new GridCacheSetHeader(IgniteUuid.randomUuid(), collocated);
+            if (!collocated) {
+                // First search for old version.
+                hdr = (GridCacheSetHeader)cache.get(key);
 
-                GridCacheSetHeader old = (GridCacheSetHeader)cache.getAndPutIfAbsent(key, hdr);
+                // If old version was not found than create new header but don't put it into cache.
+                if (hdr == null) {
+                    hdr = new GridCacheSetHeader(IgniteUuid.randomUuid(), false);
 
-                if (old != null)
-                    hdr = old;
+                    oldVer = false;
+                }
             }
-            else
-                hdr = (GridCacheSetHeader)cache.get(key);
+            else {
+                if (create) {
```

Comment:
may be collapse statements "else {  if(create) { ....."  ->  "else if(create) {....."  ?